crates/zksync/compiler/src/zksolc/compile.rs

```diff
@@ -401,40 +540,95 @@ impl ZkSolc {
         }
 
         let mut result = ProjectCompileOutput::default();
-        result.set_compiled_artifacts(Artifacts(data));
-        Ok((result, contract_bytecodes))
+        result.set_compiled_artifacts(Artifacts(all_artifacts));
+        Ok(result)
     }
 
-    /// Checks if the contract has already been compiled for the given input contract hash.
-    /// If yes, returns the pre-compiled data.
-    fn check_cache(
+    /// Converts the artifacts into canonical foundry artifact paths using it's custom
+    /// name-resolution scheme. This mirrors the artifact paths generated via solc compilation.
+    fn canonicalize_artifact_paths(
         &self,
-        artifact_paths: &ZkSolcArtifactPaths,
-        contract_hash: &str,
-    ) -> Option<Vec<u8>> {
-        if artifact_paths.contract_hash.exists() && artifact_paths.artifact.exists() {
-            File::open(&artifact_paths.contract_hash)
-                .and_then(|mut file| {
-                    let mut cached_contract_hash = String::new();
-                    file.read_to_string(&mut cached_contract_hash).map(|_| cached_contract_hash)
-                })
-                .and_then(|cached_contract_hash| {
-                    if cached_contract_hash == contract_hash {
-                        Ok(Some(contract_hash))
-                    } else {
-                        Err(std::io::Error::new(std::io::ErrorKind::Other, "hashes do not match"))
+        mut all_artifacts: BTreeMap<
+            String,
+            BTreeMap<String, ArtifactFile<ConfigurableContractArtifact>>,
+        >,
+    ) -> FileToContractsMap<Vec<ArtifactFile<ConfigurableContractArtifact>>> {
+        // Create a mapping of artifacts with their artifact keys. This will group duplicates
+        // together.
+        let mut mapped_contracts = HashMap::new();
+        for (contract_path, artifacts) in &mut all_artifacts {
+            for (artifact_key, artifact) in artifacts {
+                mapped_contracts
+                    .entry(artifact_key)
+                    .or_insert(vec![])
```

Comment:
nit: `or_default`

Reply:
type inference won't work for `mapped_contracts` with `or_default()`